snapcraft/internal/sources.py

```diff
@@ -235,20 +265,54 @@ def pull(self):
 
 class Tar(FileBase):
 
-    def __init__(self, source, source_dir, source_tag=None,
+    def __init__(self, source, source_checksum, source_dir, source_tag=None,
                  source_branch=None):
-        super().__init__(source, source_dir, source_tag, source_branch)
+        super().__init__(
+            source, source_checksum, source_dir, source_tag, source_branch)
         if source_tag:
             raise IncompatibleOptionsError(
                 'can\'t specify a source-tag for a tar source')
         elif source_branch:
             raise IncompatibleOptionsError(
                 'can\'t specify a source-branch for a tar source')
 
-    def provision(self, dst, clean_target=True, keep_tarball=False):
+    def pre_check_checksum(self, source_checksum, tarball):
+        if source_checksum.startswith('http'):
+            checksum = urllib.urlopen(source_checksum)
+            source_checksum = checksum.read()
+            self.check_checksum(self, source_checksum, tarball)
+        elif os.path.isfile(source_checksum):
+            with open(source_checksum, "r") as source_file:
+                source_checksum = str(source_file.read()).rstrip()
+            self.check_checksum(self, source_checksum, tarball)
+        elif len(source_checksum) == 32 or len(source_checksum) == 64 or \
+                len(source_checksum) == 128:
+            self.check_checksum(self, source_checksum, tarball)
+
+    def check_checksum(self, source_checksum, tarball):
+        if len(source_checksum) == 32:
+            md5 = hashlib.md5(open(tarball, 'rb').read()).hexdigest()
+            if md5 != source_checksum:
+                raise NonMatchingChecksum(
+                    "the checksum doesn't match the downloaded file")
+        elif len(source_checksum) == 64:
+            sha256 = hashlib.sha256(open(tarball, 'rb').read()).hexdigest()
+            if sha256 != source_checksum:
+                raise NonMatchingChecksum(
+                    "the checksum doesn't match the downloaded file")
+        elif len(source_checksum) == 128:
+            sha512 = hashlib.sha512(open(tarball, 'rb').read()).hexdigest()
+            if sha512 != source_checksum:
+                raise NonMatchingChecksum(
+                    "the checksum doesn't match the downloaded file")
+
+    def provision(self, dst, source_checksum, clean_target=True,
+                  keep_tarball=False):
         # TODO add unit tests.
         tarball = os.path.join(self.source_dir, os.path.basename(self.source))
 
+        self.pre_check_checksum(self, source_checksum, tarball)
```

Comment:
which is https://github.com/snapcore/snapcraft/pull/619/files#diff-e6205a80af76b4faec1d9241dc3b3b7cR119